### PR TITLE
PPO Tutorial

### PR DESCRIPTION
**Pull request by vmoens**
Introduces a PPO tutorial using TorchRL.

### CODE REVIEW

File: intermediate_source/reinforcement_ppo.py [hunk 1]
```diff
@@ -0,0 +1,676 @@
...
+print("rollout of three steps:", rollout)
+print("Shape of the rollout TensorDict:", rollout.sub_batch_size)
+
+######################################################################
+# Our rollout data has a shape of ``torch.Size([3])`, which matches the number of steps
+# we ran it for. The ``"next"`` entry points to the data coming after the current step.
+# In most cases, the ``"next""`` data at time `t` matches the data at ``t+1``, but this
+# may not be the case if we are using some specific transformations (e.g. mutli-step).
+#
+# Policy
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:332]:
```suggestion
# Define a Policy
```

File: intermediate_source/reinforcement_ppo.py [hunk 2]
```diff
@@ -0,0 +1,676 @@
...
+# ------
+#
+# PPO utilizes a stochastic policy to handle exploration. This means that our
+# neural network will have to output the parameters of a distribution, rather
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:336]:
```suggestion
# neural network will have to output the parameters of a distribution rather
```

File: intermediate_source/reinforcement_ppo.py [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+#     f_{\theta}(\text{observation}) = \mu_{\theta}(\text{observation}), \sigma^{+}_{\theta}(\text{observation})
+#
+# The only extra-difficulty that is brought up here is to split our output in two
+# equal parts and map the second to a scrictly positive space.
+#
+# We design the policy in three steps:
+#
+# 1. Define a neural network ``D_obs`` -> ``2 * D_action``. Indeed, our ``loc`` (mu) and ``scale`` (sigma) both have dimension ``D_action``;
+#
+# 2. Append a :class:`NormalParamExtractor` to extract a location and a scale (ie splits the input in two equal parts
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:356]:
```suggestion
# 2. Append a :class:`NormalParamExtractor` to extract a location and a scale (for example, splits the input in two equal parts
```

File: intermediate_source/reinforcement_ppo.py [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+    nn.LazyLinear(num_cells, device=device),
+    nn.Tanh(),
+    nn.LazyLinear(num_cells, device=device),
+    nn.Tanh(),
+    nn.LazyLinear(2 * env.action_spec.shape[-1], device=device),
+    NormalParamExtractor(),
+)
+
+######################################################################
+# To enable the policy to "talk" with the environment through the tensordict
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:374]:
```suggestion
# To enable the policy to *talk* with the environment through the ``TensorDict``
```

File: intermediate_source/reinforcement_ppo.py [hunk 5]
```diff
@@ -0,0 +1,676 @@
...
+# data carrier, we wrap the ``nn.Module`` in a :class:`TensorDictModule`. This
+# class will simply ready the ``in_keys`` it is provided with and write the
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:376]:
```suggestion
# class will simply prepare the ``in_keys`` it is provided with and write the
```

File: intermediate_source/reinforcement_ppo.py [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+    in_keys=["loc", "scale"],
+    distribution_class=TanhNormal,
+    distribution_kwargs={"min": env.action_spec.space.minimum,
+                         "max": env.action_spec.space.maximum},
+    return_log_prob=True,
+    # we'll need the log-prob for the numerator of the importance weights
+)
+
+######################################################################
+# Value network
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:411]: Create Value Network?

File: intermediate_source/reinforcement_ppo.py [hunk 7]
```diff
@@ -0,0 +1,676 @@
...
+    nn.LazyLinear(1, device=device),
+)
+
+value_module = ValueOperator(
+    module=value_net,
+    in_keys=["observation"],
+)
+
+######################################################################
+# let's try our policy and value modules. As we said earlier, the usage of
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:438]:
```suggestion
# Let's try our policy and value modules. As we said earlier, the usage of
```

File: intermediate_source/reinforcement_ppo.py [hunk 8]
```diff
@@ -0,0 +1,676 @@
...
+# :class:`TensorDictModule` makes it possible to directly read the output
+# of the environment to run these modules, as they know what information to read
+# and where to write it:
+#
+print("Running policy:", policy_module(env.reset()))
+print("Running value:", value_module(env.reset()))
+
+######################################################################
+# Data collector
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:447]: Is this a heading 3?

File: intermediate_source/reinforcement_ppo.py [hunk 9]
```diff
@@ -0,0 +1,676 @@
...
+    env,
+    policy_module,
+    frames_per_batch=frames_per_batch,
+    total_frames=total_frames,
+    split_trajs=False,
+    device=device,
+)
+
+######################################################################
+# Replay buffer
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:486]:
```suggestion
# Create Replay Buffer
```

File: intermediate_source/reinforcement_ppo.py [hunk 10]
```diff
@@ -0,0 +1,676 @@
...
+#
+# Replay buffers are a common building piece of off-policy RL algorithms.
+# In on-policy contexts, a replay buffer is refilled every time a batch of
+# data is collected, and its data is repeatedly consumed for a certain number
+# of epochs.
+#
+# TorchRL's replay buffers are built using a common container
+# :class:`ReplayBuffer` which takes as argument the components of the buffer:
+# a storage, a writer, a sampler and possibly some transforms. Only the
+# storage (which indicates the replay buffer capacity) is mandatory. We
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:497]:
```suggestion
# storage replay buffer which indicates the capacity is mandatory. We
```

File: intermediate_source/reinforcement_ppo.py [hunk 11]
```diff
@@ -0,0 +1,676 @@
...
+    storage=LazyTensorStorage(frames_per_batch),
+    sampler=SamplerWithoutReplacement(),
+)
+
+######################################################################
+# Loss function
+# -------------
+#
+# The PPO loss can be directly imported from torchrl for convenience using the
+# :class:`ClipPPOLoss` class. This is the easiest way of utilising PPO:
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:515]:
```suggestion
# :class:`ClipPPOLoss` class. This is the easiest way of utilizing PPO:
```

File: intermediate_source/reinforcement_ppo.py [hunk 12]
```diff
@@ -0,0 +1,676 @@
...
+# it hides away the mathematical operations of PPO and the control flow that
+# goes with it.
+#
+# PPO requires some "advantage estimation" to be computed. In short, an advantage
+# is a value that reflects an expectancy over the return value while dealing with
+# the bias / variance tradeoff.
+# To compute the advantage, one just needs to (1) build the advantage module, which
+# utilises our value operator, and (2) pass each batch of data through it before each
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:523]:
```suggestion
# utilizes our value operator, and (2) pass each batch of data through it before each
```

File: intermediate_source/reinforcement_ppo.py [hunk 13]
```diff
@@ -0,0 +1,676 @@
...
+    optim,
+    total_frames // frames_per_batch,
+    0.0
+)
+
+######################################################################
+# Training loop
+# -------------
+# We now have all the pieces needed to code our training loop.
+# The steps are quite easy: collect data, compute advantage, loop over the collected
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:563]:
```suggestion
# The steps include:
#
# * Collect data
# * Compute advantage
# * Loop over the collected to compute loss values
# * Back propagate
# * Optimize
# * Repeat
```

File: intermediate_source/reinforcement_ppo.py [hunk 14]
```diff
@@ -0,0 +1,676 @@
...
+# data to compute loss values, backpropagate, optimize and repeat.
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:564]:
```suggestion
```

File: intermediate_source/reinforcement_ppo.py [hunk 15]
```diff
@@ -0,0 +1,676 @@
...
+plt.subplot(2, 2, 3)
+plt.plot(logs["eval reward (sum)"])
+plt.title("Return (test)")
+plt.subplot(2, 2, 4)
+plt.plot(logs["eval step_count"])
+plt.title("Max step count (test)")
+plt.show()
+
+######################################################################
+# Next steps
```
Comment (svekars) [inline intermediate_source/reinforcement_ppo.py:665]: We need a Conclusion section. I think you can rewrite this section a bit to become the conclusion section. For example:
```
In this tutorial, we have learned:
1.
2.
3.

If you want to experiment with this tutorial a bit more, you can apply the following modifications: 

* This algorithm ...

* From a logging perspective .....

For more information about TorchRL, go to <link to the TorchRL docs>

**Review by svekars — COMMENTED** [review]
Some editorial suggestions. Let me know if you have any questions.

**Comment by vmoens** [comment]
@svekars this is ready for review

**Pull request merged**